cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll

```diff
@@ -219,18 +222,36 @@ class DefImpl extends DefOrUseImpl, TDefImpl {
   }
 
   predicate isCertain() { isDef(true, _, address, _, _, ind) }
+
+  /**
+   * Holds if this definition not having a use does not imply that it
+   * will be removed from the SSA def-use relation.
+   */
+  predicate cannotBePruned() { none() }
+
+  /** Gets the definition from the SSA pruning stage that corresponds to this definition. */
+  SsaInternals0::Def getPruningDef() { result.getAddressOperand() = address }
+}
+
+private class DirectDef extends DefImpl, TDirectDef {
```

Comment:
What is the reason for splitting this off from `DefImpl`?

Reply:
Ah, 86db73f3ff69c3d35f1707a6fb23aa6e23485a44 is the reason.

Reply:
Yeah, exactly. Sorry I didn't get back to you on that in time 😄